pkg/controller/master/image/vm_image_controller.go

```diff
@@ -57,6 +57,18 @@ func (h *vmImageHandler) OnChanged(_ string, image *harvesterv1.VirtualMachineIm
 			return h.images.Update(toUpdate)
 		}
 
+		// sync virtualSize (handles the case for existing images that were
+		// imported before this field was added, because adding the field to
+		// the CRD triggers vmImageHandler.OnChanged)
+		if image.Status.VirtualSize == 0 {
+			bi, err := util.GetBackingImage(h.backingImageCache, image)
+			if err == nil {
+				toUpdate := image.DeepCopy()
+				toUpdate.Status.VirtualSize = bi.Status.VirtualSize
+				return h.images.Update(toUpdate)
+			}
```

Comment:
If we get an error from `util.GetBackingImage(h.backingImageCache, image)`, but do not return the error, it will make the image not be reconciled again, and be left with an empty `VirtualSize`

Reply:
Ah, right. Maybe best I move this block into `processVMImage()` after that function's call to `util.GetBackingImage()` to handle that case, and avoid a bit of code duplication.

Reply:
`processVMImage()` is for images not imported yet, IIUC, this is for handling images that are already imported for the upgrade path?

Reply:
My reading is that `processVMImage()` is effectively a continuation of `OnChanged()`, and was split off the function as part of https://github.com/harvester/harvester/pull/5498. If I leave the virtualSize update in `OnChanged()`, and want to handle the error from `util.GetBackingImage()`, I have to duplicate the error handling code that's already in `processVmImage()`, for example:

```diff --git a/pkg/controller/master/image/vm_image_controller.go --- a/pkg/controller/master/image/vm_image_controller.go
+++ b/pkg/controller/master/image/vm_image_controller.go
@@ -57,6 +57,26 @@ func (h *vmImageHandler) OnChanged(_ string, image *harvesterv1.VirtualMachineIm
                        return h.images.Update(toUpdate)
                }
 
+               // sync virtualSize (handles the case for existing images that were
+               // imported before this field was added, because adding the field to
+               // the CRD triggers vmImageHandler.OnChanged)
+               if image.Status.VirtualSize == 0 {
+                       bi, err := util.GetBackingImage(h.backingImageCache, image)
+                       if err != nil {
+                               if !errors.IsNotFound(err) {
+                                       logrus.WithError(err).WithFields(logrus.Fields{
+                                               "namespace": image.Namespace,
+                                               "name":      image.Name,
+                                       }).Error("failed to get backing image for vmimage")
+                                       return image, err
+                               }
+                               return h.handleRetry(image)
+                       }
+                       toUpdate := image.DeepCopy()
+                       toUpdate.Status.VirtualSize = bi.Status.VirtualSize
+                       return h.images.Update(toUpdate)
+               }
+
                return image, nil
        }
 ```

Whereas if I move the virtualSize update to `processVMImage()`, I should only have to do this:

```diff --git a/pkg/controller/master/image/vm_image_controller.go b/pkg/controller/master/image/vm_image_controller.go
index 5f832ccf..06ec076e 100644
--- a/pkg/controller/master/image/vm_image_controller.go
+++ b/pkg/controller/master/image/vm_image_controller.go
@@ -88,6 +88,17 @@ func (h *vmImageHandler) processVMImage(image *harvesterv1.VirtualMachineImage)
                return h.handleRetry(image)
        }
 
+       if harvesterv1.ImageImported.IsTrue(image) {
+               // sync virtualSize (handles the case for existing images that were
+               // imported before this field was added, because adding the field to
+               // the CRD triggers vmImageHandler.OnChanged)
+               if image.Status.VirtualSize == 0 {
+                       toUpdate := image.DeepCopy()
+                       toUpdate.Status.VirtualSize = bi.Status.VirtualSize
+                       return h.images.Update(toUpdate)
+               }
+       }
+
        if bi.DeletionTimestamp != nil {
                h.imageController.EnqueueAfter(image.Namespace, image.Name, checkInterval)
                return image, nil
```

Reply:
```
func (h *vmImageHandler) OnChanged(_ string, image *harvesterv1.VirtualMachineImage) (*harvesterv1.VirtualMachineImage, error) {
       ....

	if harvesterv1.ImageImported.IsTrue(image) {
		....
		return image, nil
	}

	return h.processVMImage(image)
}
```

If the image with `harvesterv1.ImageImported.IsTrue(image) `, it's not going to enter `processVMImage()`. Or do I miss anything?

Reply:
> To avoid duplicating the util.GetBackingImage() error handling, I'd need to move that entire if harvesterv1.ImageImported.IsTrue(image) block into processVMImage.

I'm wondering this will introduce too much modification. Actually in the current implementation, we do not update `image.Status.VirtualSize` in `processVMImage()` (we update `image.Status.VirtualSize` in 'func (h *backingImageHandler) OnChanged()'), right?

IMHO, maybe we could apply the approach you mentioned before

```diff --git a/pkg/controller/master/image/vm_image_controller.go --- a/pkg/controller/master/image/vm_image_controller.go
+++ b/pkg/controller/master/image/vm_image_controller.go
@@ -57,6 +57,26 @@ func (h *vmImageHandler) OnChanged(_ string, image *harvesterv1.VirtualMachineIm
                        return h.images.Update(toUpdate)
                }
 
+               // sync virtualSize (handles the case for existing images that were
+               // imported before this field was added, because adding the field to
+               // the CRD triggers vmImageHandler.OnChanged)
+               if image.Status.VirtualSize == 0 {
+                       bi, err := util.GetBackingImage(h.backingImageCache, image)
+                       if err != nil {
+                               if !errors.IsNotFound(err) {
+                                       logrus.WithError(err).WithFields(logrus.Fields{
+                                               "namespace": image.Namespace,
+                                               "name":      image.Name,
+                                       }).Error("failed to get backing image for vmimage")
+                                       return image, err
+                               }
+                               return h.handleRetry(image)
+                       }
+                       toUpdate := image.DeepCopy()
+                       toUpdate.Status.VirtualSize = bi.Status.VirtualSize
+                       return h.images.Update(toUpdate)
+               }
+
                return image, nil
        }
 ```
 
 Or we can have the feedback from @bk201 or @FrankYang0529 ? Thanks.

Reply:
> I'm wondering this will introduce too much modification. Actually in the current implementation, we do not update `image.Status.VirtualSize` in `processVMImage()` (we update `image.Status.VirtualSize` in `func (h *backingImageHandler) OnChanged()`), right?

Correct. I was only thinking to move everything to `processVMImage()` to avoid duplicating that error handling code.

(My original reason for not handling the error case was just that I thought it extremely unlikely for `util.GetBackingImage` to fail for an already imported image)

Reply:
@bk201 @FrankYang0529 WDYT of the above?

TL;DR: we've discussed three possibilities for this little chunk of code:

1. Ignore error return from `util.GetBackingImage()` when updating virtualsize of existing imported images.
2. Handle possible error return by duplicating error handling code from `processVmImage()`, inside `OnChanged()`.
3. Move existing imported image handling code from `OnChanged()` to `processVmImage()` to avoid duplicate error handling code.

Reply:
My 2 cents is option 2, it's more straightforward.
Swallowing errors is not a good idea. If an error happens here, it probably means there is a bug somewhere because a VirtualMachineimage can't get a corresponding BackingImage.

Reply:
BTW, is an existing backing image guaranteed to have a positive `bi.Status.VirtualSize` value after Longhorn is upgraded?

Reply:
> My 2 cents is option 2, it's more straightforward.

Alright, 2 it is. Thanks @bk201, will do, then a quick re-test for good measure, then squash and we're finally done :-)

> BTW, is an existing backing image guaranteed to have a positive `bi.Status.VirtualSize` value after Longhorn is upgraded?

Yep (see my manual test notes in https://github.com/harvester/harvester/pull/5387#discussion_r1654424863)